csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,246 @@
+  //
+  // TODO: Make this private
+  bool mapThroughExpr(Expr* first, Expr* second, bool forward);
+
+  // Map through loop swizzles, as input/output IterDomains are exact, only the
+  // order they're traversed differs.
+  void mapThroughLoopSwizzles();
+
+  // Maps iter domain pairs returned by calling that return mappings from
+  // IdGraph::isTrivialExpr on every expression in the graph.
+  void mapThroughTrivialExprs();
+
+  // Removes expressions from unique_definitions_ and unique_uses_ that return
+  // mappings from IdGraph::isTrivialExpr
+  void removeTrivialExprs();
+
+  // See comment on propagate_expr_ member bool for description
+  // Once disabled this can't be reenabled on a graph. If it's reenabled it's
+  // hard to predict how mappings will propagate, which will be triggered on the
+  // next mapping. To support changing this flag, we should likely run through
+  // all expressions currently registered and propagate through all of them on
+  // switch. Then once enabled it couldn't be redisabled because we don't record
+  // the history of mapId calls.
+  void disableExprPropagation() {
+    propagate_exprs_ = false;
+  }
+
+  // Removes the provided expression group from unique_definitions_ and
+  // unique_uses_ breaking traversal through them.
+  void eraseExprGroup(ExprGroup expr_group);
+
+  // Returns if the expression group has an input id group that matches an
+  // output id group. This means traversing on this expression doesn't actually
+  // do anything.
+  bool isTrivialExprGroup(ExprGroup expr_group) const;
+
+ private:
+  // If propagate_exprs_ = false, then mapThroughExpr will not be called as a
+  // consequence of calling mapIds. As well as mapThroughExpr will not be called
+  // (again) as a result of calling mapThroughExpr.
+  //
+  // Note: For the second sentence of above... mapThroughExpr can call mapIds
+  // which could in return call mapThoughExpr again, but propagate_exprs_ as
+  // mentioned above prevents that from happening.
+  //
+  // TODO: Should propagate_exprs_ be a const member?
+  bool propagate_exprs_ = true;
+
+  // Keeps a disjoint set entry for all IterDomain for all mapping mode types.
+  //
+  // Using an array here might be nice, but it seems hard to use an enum as an
+  // array key
+  // https://stackoverflow.com/questions/2102582/how-can-i-count-the-items-in-an-enum
+  DisjointSets<IterDomain*> disjoint_ids_;
+
+  // Keeps a disjoint set entry for all Expressions for all mapping mode types.
+  DisjointSets<Expr*> disjoint_exprs_;
+
+  // Hold a set of IterDomains that are considered view rfactor ids. This
+  // identification is particularly important to understand if split operations
+  // are divisible or not.
+  //
+  // TODO: This should just be in IterDomainGraphs, not here.
+  std::unordered_set<IterDomain*> view_rfactor_ids_;
+
+  std::unordered_map<IdGroup, ExprGroups> unique_definitions_;
+
+  std::unordered_map<IdGroup, ExprGroups> unique_uses_;
```

Comment:
IIUC, each `IdGroup` will have an index. Is this correct? If so, can each id group have multiple uses? It's hard to imagine that indices come down from multiple paths and all these paths get exactly the same `Val*`.